File: src/IotBbq.App/IotBbq.App/Services/Implementation/AlarmService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Windows.Devices.Gpio;

namespace IotBbq.App.Services.Implementation
{
    public class AlarmService : IAlarmService
    {
        public bool IsAlarming { get; set; }

        public event EventHandler<bool> AlarmStateChanged;

        public Timer alarmTimer;

        private GpioController gpio;

        private GpioPin pinToBuzz;

        private DateTime? whenToStop;

        private ManualResetEvent neverSignaledEvent = new ManualResetEvent(false);

        private static TimeSpan NormalPriorityInterval = TimeSpan.FromMilliseconds(300);

        private static TimeSpan HighPriorityInterval = TimeSpan.FromMilliseconds(150);

        private AlarmPriority currentPriority;

        public AlarmService()
        {
            this.alarmTimer = new Timer(this.OnAlarmTimerTick, null, Timeout.Infinite, Timeout.Infinite);
            this.gpio = GpioController.GetDefault();

            this.pinToBuzz = this.gpio.OpenPin(16, GpioSharingMode.Exclusive);
            this.pinToBuzz.SetDriveMode(GpioPinDriveMode.Output);

            // Make sure we start low
            this.pinToBuzz.Write(GpioPinValue.Low);
        }

        private void OnAlarmTimerTick(object state)
        {
            if (this.whenToStop.HasValue && DateTime.UtcNow >= this.whenToStop)
            {
                this.Silence();
                return;
            }

            this.DoOneBeep(new TimeSpan(NormalPriorityInterval.Ticks / 2));
        }

        public void Silence()
        {
            if (this.IsAlarming)
            {
                // Turn off the alarm timer
                this.alarmTimer.Change(Timeout.Infinite, Timeout.Infinite);

                this.IsAlarming = false;
                this.AlarmStateChanged?.Invoke(this, false);
            }
        }

        public void TriggerAlarm(AlarmPriority priority, TimeSpan? duration = null)
        {
            if (!this.IsAlarming || (priority == AlarmPriority.High && this.currentPriority == AlarmPriority.Normal))
            {
                if (duration.HasValue)
                {
                    this.whenToStop = DateTime.UtcNow + duration.Value;
                }
                else
                {
                    this.whenToStop = null;
                }

                // Trigger the timer immediately
                this.currentPriority = priority;
                this.alarmTimer.Change(
                    TimeSpan.Zero,
                    priority == AlarmPriority.High ? HighPriorityInterval : NormalPriorityInterval);

                this.IsAlarming = true;
                this.AlarmStateChanged?.Invoke(this, true);
            }
        }

        private void DoOneBeep(TimeSpan singleBeepDuration)
        {
            this.pinToBuzz.Write(GpioPinValue.High);
            this.neverSignaledEvent.WaitOne(singleBeepDuration);
            this.pinToBuzz.Write(GpioPinValue.Low);
        }
    }
}

File: src/IotBbq.App/IotBbq.JsApp/src/app/services/AlarmService.ts

import { IAlarmService, AlarmPriority } from './IAlarmService';
import { Utility } from './Utility';
import { Observable, Subscription, timer } from 'rxjs';
import { Inject } from '@angular/core';
import { GPIO_FACTORY_TOKEN, IGpio, IGpioFactory, InOrOut, PinValue } from './IGpio';

export class AlarmService implements IAlarmService {

  public alarmStateChanged: (state: boolean) => void = null;

  private _isAlarming = false;

  private whenToStop: Date = null;

  private normalPriorityInterval = 300;

  private highPriorityInterval = 150;

  private currentPriority: AlarmPriority = AlarmPriority.Normal;

  private alarmTimer: Observable<number> = null;

  private alarmSubscription: Subscription;

  private gpioPin: IGpio;

  private pinNumber = 16;

  constructor(
    @Inject(GPIO_FACTORY_TOKEN) gpioFactory: IGpioFactory,
  ) {
    this.gpioPin = gpioFactory.open(this.pinNumber, InOrOut.Out);
  }

  public isAlarming(): boolean {
    return this._isAlarming;
  }

  public silence(): void {
    if (this.isAlarming()) {
      this.alarmSubscription.unsubscribe();
      this.alarmTimer = null;
      this._isAlarming = false;
      const cb = this.alarmStateChanged;
      if (cb) {
        cb(false);
      }
    }
  }

  public triggerAlarm(priority: AlarmPriority, duration: number = null) {
    if (!this.isAlarming() || (priority === AlarmPriority.High && this.currentPriority === AlarmPriority.Normal)) {
      const now = new Date();
      if (duration) {
        // TODO: confirm now.getTime() is in ms
        this.whenToStop = new Date(now.getTime() + duration);
      } else {
        this.whenToStop = null;
      }

      // Trigger the timer immediately
      this.currentPriority = priority;
      this._isAlarming = true;
      this.alarmTimer = timer(0, priority === AlarmPriority.High ? this.highPriorityInterval : this.normalPriorityInterval);
      this.alarmSubscription = this.alarmTimer.subscribe(async () => await this.onAlarmTimerTick());

      const cb = this.alarmStateChanged;
      if (cb) {
        cb(true);
      }
    }
  }

  protected async doOneBeep(durationInMs: number): Promise<void> {
    this.gpioPin.write(PinValue.High);
    await Utility.sleep(durationInMs);
    this.gpioPin.write(PinValue.Low);
  }

  private async onAlarmTimerTick() {
    const now = new Date();
    if (this.whenToStop && now >= this.whenToStop) {
      this.silence();
      return;
    }

    // I think this is divided by 2 so there is an even space
    // between the tones when it is repeated
    await this.doOneBeep(this.normalPriorityInterval / 2);
  }
}
